source/extensions/http/header_formatters/preserve_case/preserve_case_formatter.cc

```diff
@@ -0,0 +1,61 @@
+#include "envoy/extensions/http/header_formatters/preserve_case/v3/preserve_case.pb.h"
+#include "envoy/extensions/http/header_formatters/preserve_case/v3/preserve_case.pb.validate.h"
+#include "envoy/http/header_formatter.h"
+#include "envoy/registry/registry.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace Http {
+namespace HeaderFormatters {
+namespace PreserveCase {
+
+class PreserveCaseHeaderFormatter : public Envoy::Http::StatefulHeaderKeyFormatter {
+public:
+  // Envoy::Http::StatefulHeaderKeyFormatter
+  std::string format(absl::string_view key) const override {
+    const auto remembered_key = original_header_keys_.find(key);
+    if (remembered_key != original_header_keys_.end()) {
+      return *remembered_key;
+    } else {
+      return std::string(key);
+    }
+  }
+  void rememberOriginalHeaderKey(absl::string_view key) override {
+    original_header_keys_.emplace(key);
```

Comment:
Suggest commenting that this will implement a policy of latching the casing of the first instance of a header, so:
```
Cookie: foo
cookie: bar
```
will reserialize as
```
Cookie: foo
Cookie: bar
```
I think that's the right call (rather than awkwardly remembering the case of each instance) but it is a decision you might as well specify here, and also test for.

Reply:
+1 for commenting this in the API, in case someone is confused about how it may not preserve case for instances of multiple header fields

Reply:
Yeah agreed, I thought about this but got lazy. Will do.